site/student-lectures/cultural-algorithms/cultural-algorithms.rst

```diff
@@ -0,0 +1,255 @@
+
+    .. code-block:: text
+
+        {
+         'situational': None,
+         'normative': [[-10, 10, 100, 100], [-10, 10, 100, 100]]
+        }
+
+
+
+Acceptance Function
+===================
+* Pick top n individuals from population
+* This is used to update belief space
+* There are two ways of method to pick top n individuals: **static method** and **dynamic method**
+* Static method: select a certain percentage (fixed number) of individuals from the population
+* Dynamic methods: the number is variable and changes from generation to generation
+* In the minimization example, static method is used.
+
+.. literalinclude:: ca.py
+    :language: python
+    :lineno-match:
+    :pyobject: accept
+
+
+Update Function
+===============
+* Based on n individuals return from acceptance function
+* Perform updates on components of belief space
+* Situational knowledge: store the best individual
+* Normative knowledge: boundaries are widened/narrowed
+
+* The mathematical expression for updating normative knowledge is slightly complex. I will illustrate with the following example
+* For example, now I have the initialized belief space
+* Then use one of individuals :math:`(6, 5)` return from acceptance function to update the belief space:
+
+    .. code-block:: python
+
+        # individual: (6, 5), its fitness: 61
+
+        # 1. update situational knowledge
+        best = situational
+        if situational is None or fitness(individual) < fitness(best):
+            situational = individual
+        # situational is None originally, now update to (6, 5)
+
+        # 2. update normative knowledge (this part use loop, but I separate them for explanation
+
+        # 2.1 update the bound for x_1, which is [-10, 10, 100, 100], representing [min, max, Lower, Upper]
+        # individual = (6, 5)  -> individual[0] = 6
+        # bound[0] is min value for x_1, bound[2] is Lower bound for x_1 fitness value
+        # bound[1] is max value for x_1, bound[2] is Upper bound for x_1 fitness value
+
+        # 2.1.1 update min and Lower for x_1
+        if individual[0] <= bound[0] or fitness(individual) < bound[2]:
+            bound[0] = individual[0]
+            bound[2] = fitness(individual)
+        # this satisfy fitness(individual) < bound[2], which is 61 < 100
+        # so update: [-10, 10, 100, 100]  ->  [6, 10, 61, 100]
+
+        # 2.1.2 update max and Upper for x_1
+        if individual[0] >= bound[1] or fitness(individual) < bound[3]:
+            bound[0] = individual[0]
+            bound[2] = fitness(individual)
+        # this satisfy fitness(individual) < bound[2], which is 61 < 100
+        # so update: [6, 10, 61, 100]  ->  [6, 6, 61, 61]
+
+        # 2.2 update the bound for x_1, which is [-10, 10, 100, 100], representing [min, max, Lower, Upper]
+        # do same thing
+        # so x_2 is update to [5, 5, 61, 61]
+
+* Above is updated based on one individuals (6, 5). There are other individuals, repeat above process for them.
```

Comment:
be consistent with punctuation. You do not use periods in other places, so perhaps ditch it here.